### PR TITLE
Toxi bootstrap

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?

It adds Bootstrap v5 as UI kit. Each commit represents a change, so please review each to see what it's affected.

We also added one behaviour: mounting Dyno's src dir into the container so it's possible to develop quickly. Simply editing static files will do the trick for hot-reloading, but if you change the HTML files, managed by the Flask app, we need to restart the gunicorn process with `docker restart dyno`.

We found and fixed a bug in the UI, as the Delay key (`Sd`) was not properly passed to the backend.

## Why is it important?
Provide a consistent experience while using the product

## Related issues
Closes #ISSUE

## New Look&Feel

**Basic layout**
Screenshot 2021-01-14 at 13 16 54

**Showing legend in a pop-up**
Screenshot 2021-01-14 at 13 18 28

### CODE REVIEW

File: scripts/modules/proxy.py [hunk 1]
```diff
@@ -29,7 +29,10 @@ def _content(self):
             logging=None,
             healthcheck=wget_healthcheck(8000, path="/"),
             ports=["9000:8000"],
-            volumes=["/var/run/docker.sock:/var/run/docker.sock"],
+            volumes=[
+                "/var/run/docker.sock:/var/run/docker.sock",
+                "./docker/dyno:/dyno",
```
Comment (cachedout) [inline scripts/modules/proxy.py:34]: This doesn't look related to the styling changes. Could you clarify please?

Comment (mdelapenya) [reply scripts/modules/proxy.py:34]: Sure, because I need a fast development environment, I need to edit a static file (JS, HTML) and changes be updated into the deployed application, without the need of rebuilding the compose file or the image. For that reason I'm mounting the entire application SRC into the container, achieving fast reload on the static files:

- JS and CSS are automatically reloaded, so refreshing the resource in the browser will do the trick
- HTML is tricky, as it's not directly served but processed by the Flask app. I only need to restart the container (3-4 seconds) and the HTML is live after a refresh in the browser. If you know a manner to add hot-reloading to the gunicorn process, I'm up to it.

Without the above changes, I'm forced to recreate the container each time I modify a static file, and for local, fast development is not acceptable.

**Comment by mdelapenya** [comment]
@cachedout this commit 19b1f0ffc09995a233c9ce82cc89772dc2f8f595 causes the compose.py CLI to fail processing the entities to build the compose file. This is the error:

```python
Traceback (most recent call last):
  File "./scripts/compose.py", line 19, in <module>
    main()
  File "./scripts/compose.py", line 15, in main
    setup()
  File "/Users/mdelapenya/sourcecode/src/github.com/elastic/apm-integration-testing/scripts/modules/cli.py", line 192, in __call__
    self.args.func()
  File "/Users/mdelapenya/sourcecode/src/github.com/elastic/apm-integration-testing/scripts/modules/cli.py", line 586, in start_handler
    selections.add(service(**args))
  File "/Users/mdelapenya/sourcecode/src/github.com/elastic/apm-integration-testing/scripts/modules/elastic_stack.py", line 112, in __init__
    self.depends_on.append("kibana")
AttributeError: 'dict' object has no attribute 'append'
```
Reverting it fixes it.

I guess what we were discussing the other day about docker-compose versions and the compatibility with `depends_on` is related.

**Comment by cachedout** [comment]
@mdelapenya Hrm, that's very odd. Could you please rebase your PR and clean out any *.pyc files and see if this error persists?

**Comment by mdelapenya** [comment]
Yeah, I rebased and solved conflicts in my branch. It happens cloning this branch too

**Comment by cachedout** [comment]
@mdelapenya Could you please give me the exactly command you are using which produces this error? Thanks.

**Comment by mdelapenya** [comment]
```shell 
./scripts/compose.py stop && ./scripts/compose.py start --with-opbeans-python --with-opbeans-python01 --dyno master
```
I removed the generated docker-compose file too

File: docker/dyno/app/templates/index.html [hunk 2]
```diff
@@ -47,76 +46,103 @@
...
+          <div id="load-stats">
+            <div id="get_stats" class="card" id="get-stats-card"></div>
+            <div id="post_stats" class="card" id="post-stats-card"></div>
+          </div>
+      </div>
+      <div class="col-sm-10">
+        <div class="row" id="proxy-container"></div>
+      </div>
+    </div>
+    <div class="row" id="load-container"></div>
```
Comment (mdelapenya) [inline docker/dyno/app/templates/index.html:144]: I do not know what is the purpose of this div-container. Is it still needed?
```suggestion
```

File: docker/dyno/entrypoint.sh [hunk 3]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 echo "Staring server.."
-gunicorn -b 0.0.0.0 dyno.app:app --capture-output  -t 90 -w 8
+gunicorn -b 0.0.0.0 dyno.app:app --capture-output  -t 90 -w 8 --reload
```
Comment (mdelapenya) [inline docker/dyno/entrypoint.sh:3]: For same reason explained above, it allows hot-reloading the static code while editing in dev mode.

Comment (cachedout) [reply docker/dyno/entrypoint.sh:3]: 👍

**Comment by cachedout** [comment]
@mdelapenya This somehow got out of sync with my (upstream) PR. Are you certain that you did a rebase and not a merge? Either way, could you please try again? Once we get this synced up with  my latest round of changes, it should be good to merge.

File: docker/dyno/app/templates/index.html (hunk 2 above)
Comment (cachedout) [reply docker/dyno/app/templates/index.html:144]: This is necessary. Please don't remove it. It contains the `load-controls` div.

File: scripts/modules/proxy.py (hunk 1 above)
Comment (cachedout) [reply scripts/modules/proxy.py:34]: 👍 Sounds good. Thanks!

**Comment by mdelapenya** [comment]
Yeah, I had to merge `https://github.com/cachedout/apm-integration-testing/pull/1/commits/1d0a01752c153e75ed76a97fe6bd1f9ead6ffdc5` after a push-force.

Let me fetch again and resolve conflicts

**Comment by cachedout** [comment]
You should be able to do a rebase and just rewrite the whole history from the tip of my branch and it should (hopefully) do the right thing.

**Comment by mdelapenya** [comment]
Ok, I'll rebase and remove all merge-commit. Then I'll push-force this branch

**Comment by cachedout** [comment]
@mdelapenya This looks correct now. Could you please verify that it's working for you as expected? Then I'll merge it. Thanks!

**Comment by mdelapenya** [comment]
Let me return back and will do, thanks!

**Comment by mdelapenya** [comment]
I think my review is done. Let me add a few UX comments:

## Sliders
- When using the sliders, and showing their values in console.logs, I found it confusing having some of them being max value (100) at the top and others having it at the bottom. I.e. Errors and workers slider have max value at the top, but the rest have it at the bottom.
- I guess there is a reason why the min value is sometimes 0 and sometimes 1. I.e, for workers it's 0.1, errors is 1, and the rest is 0.
- Range values: I verified that sliders having 100 as maximum, go from 0 to 99, while Errors sliders go from 1 to 100.
- Should the molotov sliders (errors and workers) be disabled for the specific service if the switch for the control-panel service is disabled? Otherwise, I see network errors in console devtools when those two sliders are used and the switch for the service is OFF. On the contrary, they work again when used if the switch is ON. This behavior can be reproduced without any of my changes.
- I'd see convenient providing the current value of the slider without needing the terminal, something like a label in the slider. For the same reason, displaying its min/max values would help in getting the context of each slider.

## Services
- Not sure why redis and postgres services are present in the right panel (proxy-container), as they do not appear in the load controls panel. How do the interactions in these services' sliders affect the application?
- I'd like to provide a descriptive tittle/header to the left panel containing the services. What would be a good one?

## HTML, JS and layout
- I have not seen any reference in the code to the `load-container` div. It seems an empty div that would receive an AJAX request with content, but cannot see any place in the code using that as target.
- The sliders' UI is based in spans having an `id` with the code/metric to be modified (i.e. CPU, E, W). But it's not a best practice to have non-unique IDs across the page. I have tried to change it, prepending the serviceName, but it seems that `id` is used in the JS functions used to load the current state of the sliders. I've seen a _dangerous_ `this` keyword which prevented me to send that commit just in case I break something.

File: docker/dyno/app/__init__.py [hunk 4]
```diff
@@ -20,6 +20,11 @@ def create_app(config_class=Cfg):
 app = create_app()
 
 
+@app.after_request
```
Comment (mdelapenya) [inline docker/dyno/app/__init__.py]: This is super stupid change, suggested by lighthouse to cache static resources in the browser

**Comment by cachedout** [comment]
Hi @mdelapenya !

Thanks for this excellent feedback. I'll address each point below:

> When using the sliders, and showing their values in console.logs, I found it confusing having some of them being max value (100) at the top and others having it at the bottom. I.e. Errors and workers slider have max value at the top, but the rest have it at the bottom.

Yeah. I've really gone back and forth on this. There's no easy answer here. For example, imagine that you set all to zero by default. When that happens, the mental model gets confusing for a different because raising the slider for worker will increase the load but increasing the slider for CPU will actually _decrease_ the amount of available CPU, which I find very counter-intuitive. Potentially, we could solve this by naming things more clearly, like "CPU pressure" and "Worker pressure" instead of just "CPU" and "Load", but I'm honestly not sure. What I have now is the compromise that I personally found most closely resembled natural behavior for me but I definitely agree that it's not ideal.

> I guess there is a reason why the min value is sometimes 0 and sometimes 1. I.e, for workers it's 0.1, errors is 1, and the rest is 0.

Well-spotted! This is a side-effect of the way that normalization is done. I don't consider it problematic because we're really just trying to approximate a minimal value and a maximal value. In a practical sense, for example, it doesn't really matter much if the amount of CPU that's allocated is 99.9% or 100%. A little bit of slop is expected behavior. To avoid this, we'd have to make the math that does the normalization more complex than I would like and I don't think that for this application that trade-off is worth the cost.

> Range values: I verified that sliders having 100 as maximum, go from 0 to 99, while Errors sliders go from 1 to 100.

Strictly speaking, this is probably a very small bug but its effect on the overall application is probably undetectable. See the above answer for details. I don't think we should try to correct this unless we can show that it's problematic, because, again, the cost in complexity may not make it worth it.

> Should the Molotov sliders (errors and workers) be disabled for the specific service if the switch for the control-panel service is disabled? Otherwise, I see network errors in console dev-tools when those two sliders are used and the switch for the service is OFF. On the contrary, they work again when used if the switch is ON. This behavior can be reproduced without any of my changes.

Yes! I agree that we should do this. I don't think that we should block this PR for it, but the concept of "disabling" a slider is something I definitely want to do in the near future. This should also be done to disable sliders which have no effect, such as  the "Workers" slider for the Postgres and Redis instances. To do this properly requires a bit more logic to be added to the API layer, which is why I've been putting it off. I plan to do this in an upcoming round of improvements.

> I'd see convenient providing the current value of the slider without needing the terminal, something like a label in the slider. For the same reason, displaying its min/max values would help in getting the context of each slider.

Unfortunately, I disagree here. I had this in the application for a while and it was not helpful at all, and even very distracting at times. The reason for this again is that the value on the slider is normalized to a raw value that's passed to the various service, using a range of values (which are declared in range.yml). Even if we were to add a "50" at the top of each slider to indicate that you were in the middle of the range, I don't think that's helpful to the user because it doesn't give them actionable information. All it does is to encourage them to ask "50%...of what?". We don't want them to think about the application like that. Instead, we want them to think about it in terms of relative terms like, "OK, I've added a bit of load...now a bit more...now a bit more." The _amount_ of load doesn't really matter that much. All that matters is an increase or decrease made in the Dyno control panel and the corresponding effect in the APM application. If we encourage people to think in terms of raw values, this will all get very messy and moves us further away from what the application does well. Applying pressure to applications in precise amounts is a very _different_ problem than the one this application intends to solve, and we shouldn't encourage users to confuse the two.

> Not sure why redis and postgres services are present in the right panel (proxy-container), as they do not appear in the load controls panel. How do the interactions in these services' sliders affect the application?

Great question! Let's take the `opbeans-python` application as an example. It's a Django application, but the database it uses is backed by the Postgres instance. It also passes messages to and from the Redis instance. Interaction with both Postgres and Redis is observed via the APM application. Therefore, if, for example, you lower the amount of CPU on the Postgres instance, you should see a decrease in the request-rate capacity of the Opbeans application. You may have noticed that the "W" and "E" sliders have no effect Postgres or Redis (and that they are not recognized by the load-generation pane). This is a known issue and I discuss disabling those sliders in the answer to a previous question above.

> I'd like to provide a descriptive tittle/header to the left panel containing the services. What would be a good one?

If you're referring to the list of applications to which load can be applied on the left-hand side of the page, perhaps just "Opbeans" would be a good title? Another good one would be "Applications".

> I have not seen any reference in the code to the load-container div. It seems an empty div that would receive an AJAX request with content, but cannot see any place in the code using that as target.

If it can be safely removed without changing the way the application renders on the page, then I'm fine with removing it.

> The sliders' UI is based in spans having an id with the code/metric to be modified (i.e. CPU, E, W). But it's not a best practice to have non-unique IDs across the page. I have tried to change it, prepending the serviceName, but it seems that id is used in the JS functions used to load the current state of the sliders. I've seen a dangerous this keyword which prevented me to send that commit just in case I break something.

I agree that this is probably not the best design. In the interest of trying to get this PR merged, could we address that once we have some better test coverage in on the JS side?

File: docker/dyno/app/__init__.py (hunk 4 above)
Comment (cachedout) [reply docker/dyno/app/__init__.py]: I don't understand why we would add this? We're already sending back HTTP 304s and Cache-Control headers are already being set correctly.
Capture d’écran 2021-01-20 à 10 44 45
Capture d’écran 2021-01-20 à 10 45 47

Comment (mdelapenya) [reply docker/dyno/app/__init__.py]: I'm ok with reverting it :)

**Comment by mdelapenya** [comment]
Hi @cachedout, I agree with your answers and I'm not opposed to any of the suggested changes, which I've already sent. Let's do another improvements round after merging this one.

Thank you!

**Comment by mdelapenya** [comment]
@cachedout are we planning to include these UI refinements any time soon? I saw a backport for Dyno elastic/apm-integration-testing#1204 not including the `toxi` branch. Are we missing to merge that branch?
